tpm2/marshalling.go

```diff
@@ -22,6 +23,45 @@ type marshallableWithHint interface {
 	get(hint int64) (reflect.Value, error)
 }
 
+// CommandPreimage represents a structured preimage of cpHash for a TPM command.
+// This structure is marshaled to bytes using [Marshal] for storage/transmission
+// and can be converted to the raw cpHash preimage format for hashing.
+//
+// Format when marshaled:
+//   - CommandCode: 4 bytes (TPMCC)
+//   - Names: sized list of TPM2BName
+//   - Parameters: sized buffer
+//
+// See definition in Part 1: Architecture, section 16.7.
+type CommandPreimage struct {
+	marshalByReflection
+	// CommandCode is the TPM command code
+	CommandCode TPMCC
+	// Names are the names of the handles referenced by the command
+	Names []TPM2BName `gotpm:"list"`
+	// Parameters are the marshaled command parameters
+	Parameters TPM2BData
+}
+
+// ToCPHashPreimage converts the CommandPreimage to the raw buffer format
+// used to compute cpHash according to TPM 2.0 spec.
+func (cp *CommandPreimage) ToCPHashPreimage() []byte {
```

Comment:
Why not use the raw preimage of cphash/rphash for the serialized format, so that a caller can directly compute those hashes from the serialized byte arrays rather than unmarshalling and remarshalling again with `ToCPHashPreimage`?

If it's a matter of encoding how many handles each command/response uses, it might be possible to write a helper that uses `cmdNames` and just counts the size.